### ISSUE TITLE
stats.lua when watching 3D. Different problems with gpu.next vs gpu

### ISSUE THREAD

**Issue by netExtra** [mpv-player/mpv#12462]
### Important Information
When watching 3D each frame is divided into 2 sub-frames (up/down or side/side).  See attached photos.

vo=gpu
When using vo=gpu, each sub-frame has a copy of stats.lua which means then when 3D is enabled they effectively join together to become one stats.lua output. But the scaled resolution is incorrect. For 1080p it correctly reports each division as being 1920x540 but when they are joined into one frame this should now read 1920x1080.

vo=gpu-next
When using vo=gpu-next, there is only one stats display across both sub-frames. Effective treating the display as one screen.  As such it correctly reports the scaled resolution as 1920x1080. But when 3d is enabled and both sub-screens join into one screen the stats become unreadable.

- mpv version - Any MPV version
- Windows Version - Windows 10
- Source of the mpv binary - Any mpv version

- Possible screenshot or video of visual glitches
<img width="574" alt="gpu stats lua" src="https://github.com/mpv-player/mpv/assets/140253706/dc760717-48af-47d4-8d1b-b7e53fee2b53">

<img width="735" alt="gpu-next stats lua" src="https://github.com/mpv-player/mpv/assets/140253706/5716b121-b5be-43c3-a0fc-9453de123a9f">

### Reproduction steps
Described in section Important Information

### Expected behavior / Prefered behaviour
Easiest option would be to set the scaled resolution when vo=gpu to the total screen size as it is currently done in vpo=gpu-next.

### Actual behavior
For vo=gpu it works better than vo=gpu-next but the scaled resolution is incorrect when the two sub-screens are combined.

### Log file
Is this required?

**Comment by netExtra** [comment]
Just need some help on where to look to "try" and fix this. I couldn't locate the relevant code.

**Comment by netExtra** [comment]
Sorry but any help or pointers so I can look into this myself?

**Comment by haasn** [comment]
I suppose the first thing to figure out would be whether this should be handled by mpv or by libplacebo. Does vo_gpu generate the OSD twice (once per eye), or does it generate the OSD once (at halved resolution) and paint it twice?

**Comment by haasn** [comment]
A quick glance at [mpgl_osd_generate](https://github.com/mpv-player/mpv/blob/master/video/out/gpu/osd.c#L329) reveals a lot of code related to stereo3D support.

Presumably that code needs to be lifted out of `vo_gpu` and into some common place where both VOs can access it. Or maybe even the common OSD code.

**Comment by netExtra** [comment]
> A quick glance at [mpgl_osd_generate](https://github.com/mpv-player/mpv/blob/master/video/out/gpu/osd.c#L329) reveals a lot of code related to stereo3D support.
> 
> Presumably that code needs to be lifted out of `vo_gpu` and into some common place where both VOs can access it. Or maybe even the common OSD code.
TBH I wasn't expecting it to look like this. I just naively assumed that stats.lua was being called. Couldn't be more wrong.

**Comment by netExtra** [comment]
Hi. I saw the comment about changes to stats.lua and I was hoping this was one of them :) Will it be possible to include this change so that the vo=gpu-next output matches the the soon to be retired vo=gpu output for 3D?
Thanks.

**Comment by netExtra** [comment]
Will the recent changes to stats.lua make it easier for the gpu-next stats output to match the vo=gpu stats output when viewed in 3D?

**Comment by llyyr** [comment]
No, and if any PR does that then it'll probably ping this issue.

**Comment by netExtra** [comment]
Thanks for the update. I guess I will just have to put up with not being able to use stats in 3D or switch back to vo=gpu which kind of defeats the purpose of having gpu-next.

**Comment by netExtra** [comment]
Will gpu-next get updated to match the 3D stats format of vo=gpu? Asking for a friend :)

**Comment by kasper93** [comment]
> Presumably that code needs to be lifted out of vo_gpu and into some common place where both VOs can access it. Or maybe even the common OSD code.

This is good exercise if someone want's to give it a try. It should be easy enough to move this code and use in vo_gpu_next.c.

**Comment by netExtra** [comment]
> > Presumably that code needs to be lifted out of vo_gpu and into some common place where both VOs can access it. Or maybe even the common OSD code.
> 
> This is good exercise if someone want's to give it a try. It should be easy enough to move this code and use in vo_gpu_next.c.

It's been a minute lol. I remember looking for the relevant vo=gpu code but I was not able to find it.

**Comment by kasper93** [comment]
@netExtra: I recall that @Arcitec tried to do that, but it turned out to be annoying. I don't remember details though. Maybe he will have more insight.

**Comment by kasper93** [comment]
Should be resolved by https://github.com/mpv-player/mpv/pull/17628 if you want to give a try.
